File: Baekjoon/1774_우주신과의 교감/B_1774.py
import math
import sys


def union(x, y):
    px = find_set(x)
    py = find_set(y)

    p[py] = px


def find_set(x):
    if p[x] == x:
        return x
    p[x] = find_set(p[x])
    return p[x]


def get_distance(first, second) -> float:
    y1, x1 = first
    y2, x2 = second

    dist = math.sqrt(abs(y1 - y2) ** 2 + abs(x1 - x2) ** 2)
    return dist


N, M = map(int, sys.stdin.readline().split())
coordinates = []
for _ in range(N):
    y, x = map(int, sys.stdin.readline().split())
    coordinates.append([y, x])

linked_routes = []
for _ in range(M):
    route1, route2 = map(int, sys.stdin.readline().split())
    linked_routes.append([route1, route2])

new_coordinates = []
for i in range(len(coordinates) - 1):
    first_coordinate = coordinates[i]
    for j in range(i + 1, len(coordinates)):
        second_coordinate = coordinates[j]
        d = get_distance(first_coordinate, second_coordinate)
        new_coordinates.append([d, i + 1, j + 1])

new_coordinates.sort()

p = [_ for _ in range(N + 1)]
for i in range(len(linked_routes)):
    r1, r2 = linked_routes[i]
    if find_set(r1) != find_set(r2):
        union(r1, r2)

answer = 0
for i in range(len(new_coordinates)):
    distance, idx1, idx2 = new_coordinates[i]

    if find_set(idx1) == find_set(idx2):
        continue

    union(idx1, idx2)
    answer += distance


print("{:.2f}".format(answer))

File: Baekjoon/1774_우주신과의 교감/B_1774.js
// 이미 연결된 우주신들의 union 연산을 수행한 후에
// 거리를 구한 정점들을 반복문으로 돌면서
// 같은 루트를 갖지 않는, 아직 연결되지 않은 정점들을 union 연산하고 값을 더해준다

const unionSet = function(x, y) {
    const px = findSet(x)
    const py = findSet(y)

    p[py] = px
}

const findSet = function(x) {
    if (p[x] === x) {
        return p[x]
    }

    p[x] = findSet(p[x])
    return p[x]
}

const getDistance = function(first, second) {
    const [firstY, firstX] = first
    const [secondY, secondX] = second

    const dist = Math.sqrt(Math.abs(firstY - secondY) ** 2 + Math.abs(firstX - secondX) ** 2, 2)
    return dist
}

const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const [N, M] = input[0].split(' ').map(v => Number(v))
let coordinates = []

for (let i = 1; i <= N; i++) {
    const coordinate = input[i].split(' ').map(v => Number(v))
    coordinates.push(coordinate)
}

let linkedIdx = []

for (let i = N + 1; i < N + 1 + M; i++) {
    const idx = input[i].split(' ').map(v => Number(v))
    linkedIdx.push(idx)
}

let p = new Array(N + 1).fill(0)
for (let i = 1; i <= N; i++) {
    p[i] = i
}

for (let i = 0; i < linkedIdx.length; i++) {
    const [a, b] = linkedIdx[i]

    unionSet(a, b)
}

let distances = []
for (let i = 0; i < coordinates.length - 1; i++) {
    const firstDistance = coordinates[i]

    for (let j = i + 1; j < coordinates.length; j++) {
        const secondDistance = coordinates[j]
        const d = getDistance(firstDistance, secondDistance)

        distances.push([i + 1, j + 1, d])
    }
}

distances.sort((a, b) => {
    return a[2] - b[2]
})

let sums = 0
for (let i = 0; i < distances.length; i++) {
    const [start, end, value] = distances[i]

    if (findSet(start) === findSet(end)) {
        continue
    }

    unionSet(start, end)
    sums += value
}

console.log(sums.toFixed(2))
